qiskit/primitives/base_estimator.py

```diff
@@ -227,12 +270,105 @@ def __call__(
             values = parameter_values[i].
 
         Args:
-            circuit_indices: the list of circuit indices.
-            observable_indices: the list of observable indices.
+            circuits: the list of circuit indices or circuit objects.
+            observables: the list of observable indices or observable objects.
             parameter_values: concrete parameters to be bound.
             run_options: runtime options used for circuit execution.
 
         Returns:
             EstimatorResult: The result of the estimator.
+
+        Raises:
+            QiskitError: For mismatch of object id.
+            QiskitError: For mismatch of length of Sequence.
         """
+
+        # Support ndarray
+        if isinstance(parameter_values, np.ndarray):
+            parameter_values = parameter_values.tolist()
+
+        # Allow objects
+        try:
+            circuits = [
+                next(_finditer(id(circuit), self._circuit_ids))
+                if not isinstance(circuit, (int, np.integer))
+                else circuit
+                for circuit in circuits
+            ]
+        except StopIteration as err:
+            raise QiskitError(
+                "The circuits passed when calling estimator is not one of the circuits used to "
+                "initialize the session."
+            ) from err
+        try:
+            observables = [
+                next(_finditer(id(observable), self._observable_ids))
+                if not isinstance(observable, (int, np.integer))
+                else observable
+                for observable in observables
+            ]
+        except StopIteration as err:
+            raise QiskitError(
+                "The observables passed when calling estimator is not one of the observables used to "
+                "initialize the session."
+            ) from err
+
+        # Allow optional
+        if parameter_values is None:
+            for i in circuits:
+                if len(self._circuits[i].parameters) != 0:
+                    raise QiskitError(
+                        f"The {i}-th circuit is parameterised,"
+                        "but parameter values are not given."
+                    )
+            parameter_values = [[]] * len(circuits)
+
+        # Validation
+        if len(circuits) != len(observables):
+            raise QiskitError(
+                f"The number of circuits ({len(circuits)}) does not match "
+                f"the number of observables ({len(observables)})."
+            )
+        if len(circuits) != len(parameter_values):
+            raise QiskitError(
+                f"The number of circuits ({len(circuits)}) does not match "
+                f"the number of parameter value sets ({len(parameter_values)})."
+            )
+
```

Comment:
```suggestion

        for i, value in zip(circuits, parameter_values):
            if len(value) != len(self._parameters[i]):
                raise QiskitError(
                    f"The number of values ({len(value)}) does not match "
                    f"the number of parameters ({len(self._parameters[i])}) for the {i}-th circuit."
                )

```
This was added in sampler but missing from estimator.